doc/subsystems/jinjagen/jinjagen.rst

```diff
@@ -0,0 +1,256 @@
+..
+    Copyright (c) 2004-2015 Ned Batchelder
+    SPDX-License-Identifier: MIT
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+    Copyright (c) 2018 Nordic Semiconductor ASA
+    SPDX-License-Identifier: Apache-2.0
+.. _jinjagen_intro:
+
+Introduction
+############
+
+For some repetitive or parameterized coding tasks, it's convenient to
+use a code generating tool to build C code fragments, instead of writing
+(or editing) that source code by hand. Such a tool can also access CMake build
+parameters and device tree information to generate source code automatically
+tailored and tuned to a specific project configuration.
+
+The Zephyr project supports a code generating tool that processes template files
+and generates output files. It can be used, for example, to generate source code 
+that creates and fills data structures, adapts programming logic, creates
+configuration-specific code fragments, and more.
+
+.. image:: template_flow.png
+   :width: 500px
+   :align: center
+   :alt: Principle
+
+The code generator uses a template engine for python called jinja2. The jinja2
+templates are typically stored in files with the file extention .jinja2 which are
+rendered into other files like c-source files. The documentation for jinja2 can 
+be found here `Jinja2 <http://jinja.pocoo.org/>`_. Please refer to this site for more
+extensive information, the information in the following sections are short
+descriptions of some of the key features in jinja2. 
+
+Jinja2 files are rendered into other files using the cmake extensions functions
+
+  * zephyr_sources_jinjagen(..)
+  * zephyr_library_sources_jinjagen(..). 
+
+The generated source files are added to the Zephyr sources. While building a project
+the jinja2 files are processed by jinja2 template engine and the generated source
+files are written to the build directory.
+
+A jinja2 template is written in a domain specific language. The language inherits
+concepts from python, so anyone familiar with python should find it easy to learn
+how to write jinja2 templates.
+
+Data object
+***********
+
+All templates have access to a data object called ``data``. Through this data
+object the template will have access to information, like the elements in
+the .config file, 
+
+.. code-block:: c
+
+	{{ data['config']['CONFIG_BOARD'] }}
+
+devicetree data,
+ 
+ .. code-block:: c
```

Comment:
don't indent this directive (same blockquote rendering problem)

Reply:
ok